### PR TITLE
vim-patch:8.0.{1356,1491,1495,1522,1538,1540},8.1.{554,670,1300,1303,1875,2377}

### PR DESCRIPTION
**Pull request by janlazo**
8.0.1491 is insufficent by itself but I had issues with the tests when merging the followup popupmnu.c patches.

I don't know what do with the other 8.0.x patches. I'm going straight  to 8.1.x patches.

### CODE REVIEW

File: runtime/autoload/provider/clipboard.vim [hunk 1]
```diff
@@ -172,6 +172,10 @@ function! s:clipboard.set(lines, regtype, reg) abort
   if jobid > 0
     call jobsend(jobid, a:lines)
     call jobclose(jobid, 'stdin')
+    if argv[0] ==# 'xclip'
```
Comment (janlazo) [inline runtime/autoload/provider/clipboard.vim]: Tested this with by copying some text via Vim's clipboard first and then running `make oldtest`.

File: src/nvim/screen.c [hunk 2]
```diff
@@ -121,7 +121,6 @@
 
 #define MB_FILLER_CHAR '<'  /* character used when a double-width character
                              * doesn't fit. */
-#define W_ENDCOL(wp)   (wp->w_wincol + wp->w_width)
 #define W_ENDROW(wp)   (wp->w_winrow + wp->w_height)
```
Comment (justinmk) [inline src/nvim/screen.c]: i would perhaps move `W_ENDROW` along with `W_ENDCOL`..

File: test/functional/ui/popupmenu_spec.lua [hunk 3]
```diff
@@ -1204,10 +1204,10 @@ describe('builtin popupmenu', function()
...
     screen:expect([[
       some long prefix before the text|
-      {1:^~                        }{n: word  }|
-      {1:~                        }{n: choice}|
-      {1:~                        }{s: text  }|
-      {1:~                        }{n: thing }|
+      {1:^~                    }{n: word      }|
+      {1:~                    }{n: choice    }|
+      {1:~                    }{s: text      }|
+      {1:~                    }{n: thing     }|
```
Comment (justinmk) [inline test/functional/ui/popupmenu_spec.lua]: why did this change? Did `PUM_DEF_WIDTH` not function in the same manner as 'pumwidth' now does?

Perhaps it's worth adding an extra test with a non-default 'pumwidth'?

Comment (janlazo) [reply test/functional/ui/popupmenu_spec.lua]: These change again in later popupmnu.c 8.0.x patches. vim-patch:8.0.1491 added new code to handle non-default 'pumwidth',

File: runtime/doc/options.txt [hunk 4]
```diff
@@ -4529,6 +4529,13 @@ A jump table for the options with a short description can be found at |Q_op|.
 			global
 	Determines the maximum number of items to show in the popup menu for
 	Insert mode completion.  When zero as much space as available is used.
+	|ins-completion-menu|.
+
+						*'pumwidth'* *'pw'*
+'pumwidth' 'pw'		number	(default 0)
```
Comment (justinmk) [inline runtime/doc/options.txt]: technically this should go after 'pumblend', these entries are alphabetical.

**Review by U953 — APPROVED** [review]

**Review by U953 — APPROVED** [review]

File: test/functional/ui/popupmenu_spec.lua [hunk 5]
```diff
@@ -1156,10 +1156,10 @@ describe('builtin popupmenu', function()
...
     screen:expect([[
       some long prefix before the ^    |
-      {1:~                        }{n: word  }|
-      {1:~                        }{n: choice}|
-      {1:~                        }{n: text  }|
-      {1:~                        }{n: thing }|
+      {n:word           }{1:                 }|
+      {n:choice         }{1:                 }|
+      {n:text           }{1:                 }|
+      {n:thing          }{1:                 }|
```
Comment (justinmk) [inline test/functional/ui/popupmenu_spec.lua]: why did the menu position move to the left?

**Comment by justinmk** [comment]
`src/nvim/testdir/test_popup.vim` doesn't test anything for us since it's using Vim-only `term_xx` calls. Would be worth adding another screen test to cover non-defualt 'pumwidth'.

File: runtime/doc/eval.txt [hunk 6]
```diff
@@ -3770,6 +3771,8 @@ feedkeys({string} [, {mode}])				*feedkeys()*
 		and "\..." notation |expr-quote|. For example,
 		feedkeys("\<CR>") simulates pressing of the <Enter> key. But
 		feedkeys('\<CR>') pushes 5 characters.
+		A special code that might be useful is <Ignore>, it exits the
+		wait for a character without doing anything.  *<Ignore>*
```
Comment (justinmk) [inline runtime/doc/eval.txt]: Do we actually support this? Is there a test for it?

Comment (janlazo) [reply runtime/doc/eval.txt]: It's for `K_IGNORE`. `git grep --files-with-matches '<Ignore>'` returns `runtime matches` so it's probably untested.

File: test/functional/ui/popupmenu_spec.lua (hunk 3 above)
Comment (janlazo) [reply test/functional/ui/popupmenu_spec.lua]: complete() `startcol`  + `&pumwidth` must be less than the screen width to display the pum at the `startcol`.

File: test/functional/ui/popupmenu_spec.lua (hunk 5 above)
Comment (janlazo) [reply test/functional/ui/popupmenu_spec.lua]: https://github.com/neovim/neovim/pull/11616#discussion_r361896362

File: runtime/doc/eval.txt (hunk 6 above)
Comment (janlazo) [reply runtime/doc/eval.txt]: I missed [8.1.1300](https://github.com/vim/vim/commit/2f10658b06bbdd8f25c4ff152266c808234cee0a)

File: runtime/doc/options.txt [hunk 7]
```diff
@@ -4547,6 +4540,19 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+			global
+	Determines the maximum number of items to show in the popup menu for
+	Insert mode completion.  When zero as much space as available is used.
+	|ins-completion-menu|.
+
+						*'pumwidth'* *'pw'*
+'pumwidth' 'pw'		number	(default 15)
+			global
+	Determines the minium width to use for the popup menu for Insert mode
+	completion.  |ins-completion-menu|.
```
Comment (justinmk) [inline runtime/doc/options.txt]: would be great to also mention your comment in the doc here:

> complete() `startcol`  + `&pumwidth` must be less than the screen width to display the pum at the `startcol`.

something like:

> If the cursor column  + 'pumwidth' exceed screen width, the popup menu is nudged to fit on the screen.

File: src/nvim/keymap.c [hunk 8]
```diff
@@ -309,6 +309,7 @@ static const struct key_name_entry {
   { K_ZERO,            "Nul" },
   { K_SNR,             "SNR" },
   { K_PLUG,            "Plug" },
+  { K_IGNORE,          "Ignore" },
```
Comment (justinmk) [inline src/nvim/keymap.c]: Ah, `K_IGNORE` already existed, but now it's exposed as `<Ignore>` keycode.

File: runtime/doc/options.txt (hunk 7 above)
Comment (justinmk) [reply runtime/doc/options.txt]: done in https://github.com/neovim/neovim/pull/11604

**Pull request merged**